packages/flutter/lib/src/material/material_state.dart

```diff
@@ -400,6 +430,59 @@ class _MaterialStateUnderlineInputBorder extends MaterialStateUnderlineInputBord
   InputBorder resolve(Set<MaterialState> states) => _resolve(states);
 }
 
+/// Defines an [InputBorder] that is also a [WidgetStateProperty].
+///
+/// This class exists to enable widgets with [InputBorder] valued properties
+/// to also accept [WidgetStateProperty] objects.
+///
+/// [WidgetStateInputBorder] should only be used with widgets that document
+/// their support, like [InputDecoration.border].
+///
+/// A [WidgetStateInputBorder] can be created by:
+///  1. Creating a class that extends [OutlineInputBorder] or [UnderlineInputBorder]
+///     and implements [WidgetStateInputBorder]. The class would also need to
+///     override the [resolve] method.
+///  2. Using [WidgetStateInputBorder.resolveWith] with a callback that
+///     resolves the input border in the given states.
+///  3. Using [WidgetStateInputBorder.fromMap] to assign a border with a [WidgetStateMap].
+abstract interface class WidgetStateInputBorder implements InputBorder, WidgetStateProperty<InputBorder> {
+  /// Creates a [WidgetStateInputBorder] using a [WidgetPropertyResolver]
+  /// callback.
+  ///
+  /// This constructor should only be used for fields that support
+  /// [WidgetStateInputBorder], such as [InputDecoration.border]
+  /// (if used as a regular [OutlinedBorder], it acts the same as
+  /// an empty `OutlineInputBorder()` constructor).
```

Comment:
Mentioning OutlineInputBorder in a class that is for the generic InputBorder type feels odd. Does this need to be specifically called out?